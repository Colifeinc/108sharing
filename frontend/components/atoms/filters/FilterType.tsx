import * as React from 'react'
import FilterLayout from 'components/layouts/FilterLayout'
import { CheckBox } from 'components/atoms'

interface IType {
  id: number | string
  name: string
}

interface IProps {
  field: string
  name: string
  types: IType[]
  setIsFilterPanelOpen?(isOpen: boolean): void
  handleOnSubmit(filterValue: any): void
  maxLabelItemCount?: number
  right?: boolean
}

const initialParams = new URLSearchParams(window.location.search)
const DEFAULT_MAX_LABEL_ITEM_COUNT = 3

const FilterType: React.FC<IProps> = ({
  field,
  name,
  types: propsTypes,
  setIsFilterPanelOpen,
  handleOnSubmit,
  maxLabelItemCount: propsMaxLabelItemCount,
  right,
}) => {
  const maxLabelItemCount = propsMaxLabelItemCount ?? DEFAULT_MAX_LABEL_ITEM_COUNT
  const getInitialValue = () =>
    initialParams.has(`${field}[]`) ? initialParams.getAll(`${field}[]`) : []
  const [isOpen, setIsOpen] = React.useState(false)
  const [types, setTypes] = React.useState(getInitialValue())
  const [labelName, setLabelName] = React.useState(name)
  const setPanelIsOpen = React.useCallback(
    (open: boolean) => {
      if (open === isOpen) return
      if (setIsFilterPanelOpen) setIsFilterPanelOpen(open)
      setIsOpen(open)
    },
    [isOpen, setIsFilterPanelOpen]
  )

  const onChangeHandler = event => {
    const value = event.target.value

    if (event.target.checked) {
      if (!types.includes(value)) {
        setTypes([...types, value])
      }
    } else {
      if (types.includes(value)) {
        setTypes(types.filter(type => type !== value))
      }
    }
  }

  React.useEffect(() => {
    if (!types.length) return setLabelName(name)

    const activeTypes = propsTypes
      .map(t => (types.includes(String(t.id)) ? t.name : null))
      .filter(t => t !== null)
    const isOmitted = activeTypes.length > maxLabelItemCount
    setLabelName(`${activeTypes.slice(0, maxLabelItemCount).join(', ')}${isOmitted ? 'â€¦' : ''}`)
  }, [types, propsTypes, maxLabelItemCount, name])

  const hasValue = () => !!types.length
  const onClear = () => {
    setTypes([])
  }
  const onApply = React.useCallback(() => {
    handleOnSubmit({ [field]: types })
    setPanelIsOpen(false)
  }, [handleOnSubmit, field, types, setPanelIsOpen])

  return (
    <FilterLayout
      hasValue={hasValue()}
      label={labelName}
      onClear={onClear}
      onApply={onApply}
      right={right}
      setPanelIsOpen={setPanelIsOpen}
    >
      <form action="">
        <div>
          {propsTypes.map(type => (
            <div className="mt-3 first:mt-0" key={type.id}>
              <CheckBox
                name={'type_ids'}
                value={String(type.id)}
                onChangeHandler={onChangeHandler}
                checked={types.includes(String(type.id))}
                label={type.name}
              />
            </div>
          ))}
        </div>
      </form>
    </FilterLayout>
  )
}

export default FilterType
