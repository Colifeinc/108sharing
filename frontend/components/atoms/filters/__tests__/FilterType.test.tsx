import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import * as React from 'react'
import '../../../../test/matchMedia.mock'
import FilterType from '../FilterType'

jest.mock('../../../../core/i18n')

describe('atoms/filters/FilterType.tsx', () => {
  const handleOnSubmit = jest.fn()
  const types = [
    {
      id: 1,
      name: 'Type 1',
    },
    {
      id: 2,
      name: 'Type 2',
    },
    {
      id: 3,
      name: 'Type 3',
    },
  ]
  const props = {
    name: 'Type',
    field: 'type',
    handleOnSubmit,
    types,
  }

  it('should matches to the snapshot', () => {
    const wrapper = shallow(<FilterType {...props} />)
    expect(toJson(wrapper)).toMatchSnapshot()
  })
})
