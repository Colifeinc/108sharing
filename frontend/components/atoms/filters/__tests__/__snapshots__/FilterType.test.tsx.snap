// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`atoms/filters/FilterType.tsx should matches to the snapshot 1`] = `
<FilterLayout
  hasValue={false}
  label="Type"
  onApply={[Function]}
  onClear={[Function]}
  setPanelIsOpen={[Function]}
>
  <form
    action=""
  >
    <div>
      <div
        className="mt-3 first:mt-0"
        key="1"
      >
        <CheckBox
          checked={false}
          label="Type 1"
          name="type_ids"
          onChangeHandler={[Function]}
          value="1"
        />
      </div>
      <div
        className="mt-3 first:mt-0"
        key="2"
      >
        <CheckBox
          checked={false}
          label="Type 2"
          name="type_ids"
          onChangeHandler={[Function]}
          value="2"
        />
      </div>
      <div
        className="mt-3 first:mt-0"
        key="3"
      >
        <CheckBox
          checked={false}
          label="Type 3"
          name="type_ids"
          onChangeHandler={[Function]}
          value="3"
        />
      </div>
    </div>
  </form>
</FilterLayout>
`;
